// @polkadot/types-known/cjs/spec/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const centrifuge_chain_js_1 = tslib_1.__importDefault(require("./centrifuge-chain.js"));
const kusama_js_1 = tslib_1.__importDefault(require("./kusama.js"));
const node_js_1 = tslib_1.__importDefault(require("./node.js"));
const node_template_js_1 = tslib_1.__importDefault(require("./node-template.js"));
const polkadot_js_1 = tslib_1.__importDefault(require("./polkadot.js"));
const rococo_js_1 = tslib_1.__importDefault(require("./rococo.js"));
const shell_js_1 = tslib_1.__importDefault(require("./shell.js"));
const statemint_js_1 = tslib_1.__importDefault(require("./statemint.js"));
const westend_js_1 = tslib_1.__importDefault(require("./westend.js"));
const typesSpec = {
    'centrifuge-chain': centrifuge_chain_js_1.default,
    kusama: kusama_js_1.default,
    node: node_js_1.default,
    'node-template': node_template_js_1.default,
    polkadot: polkadot_js_1.default,
    rococo: rococo_js_1.default,
    shell: shell_js_1.default,
    statemine: statemint_js_1.default,
    statemint: statemint_js_1.default,
    westend: westend_js_1.default,
    westmint: statemint_js_1.default
};
exports.default = typesSpec;
