// @polkadot/types-known/cjs/spec/statemint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const types_create_1 = require("@polkadot/types-create");
const util_1 = require("@polkadot/util");
const sharedTypes = {
    DispatchErrorModule: 'DispatchErrorModuleU8',
    TAssetBalance: 'u128',
    ProxyType: {
        _enum: [
            'Any',
            'NonTransfer',
            'CancelProxy',
            'Assets',
            'AssetOwner',
            'AssetManager',
            'Staking'
        ]
    },
    Weight: 'WeightV1'
};
const versioned = [
    {
        minmax: [0, 3],
        types: (0, util_1.objectSpread)({
            // Enum was modified mid-flight -
            // https://github.com/paritytech/substrate/pull/10382/files#diff-e4e016b33a82268b6208dc974eea841bad47597865a749fee2f937eb6fdf67b4R498
            DispatchError: 'DispatchErrorPre6First'
        }, sharedTypes, (0, types_create_1.mapXcmTypes)('V0'))
    },
    {
        minmax: [4, 5],
        types: (0, util_1.objectSpread)({
            // As above, see https://github.com/polkadot-js/api/issues/5301
            DispatchError: 'DispatchErrorPre6First'
        }, sharedTypes, (0, types_create_1.mapXcmTypes)('V1'))
    },
    {
        // metadata V14
        minmax: [500, undefined],
        types: {
            Weight: 'WeightV1'
        }
    }
    // ,
    // {
    //   // weight v2 introduction
    //   minmax: [9300, undefined],
    //   types: {
    //     Weight: 'WeightV2'
    //   }
    // }
];
exports.default = versioned;
