// @polkadot/types-codec/cjs/base/DoNotConstruct.js
"use strict";
var _DoNotConstruct_neverError;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DoNotConstruct = void 0;
const tslib_1 = require("tslib");
/**
 * @name DoNotConstruct
 * @description
 * An unknown type that fails on construction with the type info
 */
class DoNotConstruct {
    constructor(registry, typeName = 'DoNotConstruct') {
        _DoNotConstruct_neverError.set(this, void 0);
        this.registry = registry;
        tslib_1.__classPrivateFieldSet(this, _DoNotConstruct_neverError, new Error(`DoNotConstruct: Cannot construct unknown type ${typeName}`), "f");
        throw tslib_1.__classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    static with(typeName) {
        return class extends DoNotConstruct {
            constructor(registry) {
                super(registry, typeName);
            }
        };
    }
    /**
     * @description The length of the value when encoded as a Uint8Array
     */
    get encodedLength() {
        throw tslib_1.__classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    /**
     * @description returns a hash of the contents
     */
    get hash() {
        throw tslib_1.__classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    /**
     * @description Checks if the value is an empty value (always true)
     */
    get isEmpty() {
        throw tslib_1.__classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    eq() {
        throw tslib_1.__classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    inspect() {
        throw tslib_1.__classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    toHex() {
        throw tslib_1.__classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    toHuman() {
        throw tslib_1.__classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    toJSON() {
        throw tslib_1.__classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    toPrimitive() {
        throw tslib_1.__classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    toRawType() {
        throw tslib_1.__classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    toString() {
        throw tslib_1.__classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    toU8a() {
        throw tslib_1.__classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
}
exports.DoNotConstruct = DoNotConstruct;
_DoNotConstruct_neverError = new WeakMap();
