// @polkadot/types-codec/base/DoNotConstruct.js
var _DoNotConstruct_neverError;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
/**
 * @name DoNotConstruct
 * @description
 * An unknown type that fails on construction with the type info
 */
export class DoNotConstruct {
    constructor(registry, typeName = 'DoNotConstruct') {
        _DoNotConstruct_neverError.set(this, void 0);
        this.registry = registry;
        __classPrivateFieldSet(this, _DoNotConstruct_neverError, new Error(`DoNotConstruct: Cannot construct unknown type ${typeName}`), "f");
        throw __classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    static with(typeName) {
        return class extends DoNotConstruct {
            constructor(registry) {
                super(registry, typeName);
            }
        };
    }
    /**
     * @description The length of the value when encoded as a Uint8Array
     */
    get encodedLength() {
        throw __classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    /**
     * @description returns a hash of the contents
     */
    get hash() {
        throw __classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    /**
     * @description Checks if the value is an empty value (always true)
     */
    get isEmpty() {
        throw __classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    eq() {
        throw __classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    inspect() {
        throw __classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    toHex() {
        throw __classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    toHuman() {
        throw __classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    toJSON() {
        throw __classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    toPrimitive() {
        throw __classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    toRawType() {
        throw __classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    toString() {
        throw __classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
    toU8a() {
        throw __classPrivateFieldGet(this, _DoNotConstruct_neverError, "f");
    }
}
_DoNotConstruct_neverError = new WeakMap();
