// @polkadot/types-known/spec/rococo.js
import { mapXcmTypes } from '@polkadot/types-create';
import { objectSpread } from '@polkadot/util';
/* eslint-disable sort-keys */
const sharedTypes = {
    DispatchErrorModule: 'DispatchErrorModuleU8',
    FullIdentification: '()',
    Keys: 'SessionKeys7B',
    Weight: 'WeightV1'
};
const versioned = [
    {
        minmax: [0, 200],
        types: objectSpread({}, sharedTypes, {
            AccountInfo: 'AccountInfoWithDualRefCount',
            Address: 'AccountId',
            LookupSource: 'AccountId'
        })
    },
    {
        minmax: [201, 214],
        types: objectSpread({}, sharedTypes, {
            AccountInfo: 'AccountInfoWithDualRefCount'
        })
    },
    {
        minmax: [215, 228],
        types: objectSpread({}, sharedTypes, {
            Keys: 'SessionKeys6'
        })
    },
    {
        minmax: [229, 9099],
        types: objectSpread({}, sharedTypes, mapXcmTypes('V0'))
    },
    {
        minmax: [9100, 9105],
        types: objectSpread({}, sharedTypes, mapXcmTypes('V1'))
    },
    {
        // metadata v14
        minmax: [9106, undefined],
        types: {
            Weight: 'WeightV1'
        }
    }
    // ,
    // {
    //   // weight v2 introduction
    //   minmax: [9300, undefined],
    //   types: {
    //     Weight: 'WeightV2'
    //   }
    // }
];
export default versioned;
