// @polkadot/rpc-provider/ws/index.js
var _WsProvider_instances, _WsProvider_callCache, _WsProvider_coder, _WsProvider_endpoints, _WsProvider_headers, _WsProvider_eventemitter, _WsProvider_handlers, _WsProvider_isReadyPromise, _WsProvider_stats, _WsProvider_waitingForId, _WsProvider_autoConnectMs, _WsProvider_endpointIndex, _WsProvider_endpointStats, _WsProvider_isConnected, _WsProvider_subscriptions, _WsProvider_timeoutId, _WsProvider_websocket, _WsProvider_timeout, _WsProvider_send, _WsProvider_emit, _WsProvider_onSocketClose, _WsProvider_onSocketError, _WsProvider_onSocketMessage, _WsProvider_onSocketMessageResult, _WsProvider_onSocketMessageSubscribe, _WsProvider_onSocketOpen, _WsProvider_resubscribe, _WsProvider_timeoutHandlers;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { EventEmitter } from 'eventemitter3';
import { isChildClass, isNull, isUndefined, logger, objectSpread } from '@polkadot/util';
import { xglobal } from '@polkadot/x-global';
import { WebSocket } from '@polkadot/x-ws';
import { RpcCoder } from '../coder/index.js';
import defaults from '../defaults.js';
import { LRUCache } from '../lru.js';
import { getWSErrorString } from './errors.js';
const ALIASES = {
    chain_finalisedHead: 'chain_finalizedHead',
    chain_subscribeFinalisedHeads: 'chain_subscribeFinalizedHeads',
    chain_unsubscribeFinalisedHeads: 'chain_unsubscribeFinalizedHeads'
};
const RETRY_DELAY = 2500;
const DEFAULT_TIMEOUT_MS = 60 * 1000;
const TIMEOUT_INTERVAL = 5000;
const l = logger('api-ws');
/** @internal Clears a Record<*> of all keys, optionally with all callback on clear */
function eraseRecord(record, cb) {
    Object.keys(record).forEach((key) => {
        if (cb) {
            cb(record[key]);
        }
        delete record[key];
    });
}
/** @internal Creates a default/empty stats object */
function defaultEndpointStats() {
    return { bytesRecv: 0, bytesSent: 0, cached: 0, errors: 0, requests: 0, subscriptions: 0, timeout: 0 };
}
/**
 * # @polkadot/rpc-provider/ws
 *
 * @name WsProvider
 *
 * @description The WebSocket Provider allows sending requests using WebSocket to a WebSocket RPC server TCP port. Unlike the [[HttpProvider]], it does support subscriptions and allows listening to events such as new blocks or balance changes.
 *
 * @example
 * <BR>
 *
 * ```javascript
 * import Api from '@polkadot/api/promise';
 * import { WsProvider } from '@polkadot/rpc-provider/ws';
 *
 * const provider = new WsProvider('ws://127.0.0.1:9944');
 * const api = new Api(provider);
 * ```
 *
 * @see [[HttpProvider]]
 */
export class WsProvider {
    /**
     * @param {string | string[]}  endpoint    The endpoint url. Usually `ws://ip:9944` or `wss://ip:9944`, may provide an array of endpoint strings.
     * @param {number | false} autoConnectMs Whether to connect automatically or not (default). Provided value is used as a delay between retries.
     * @param {Record<string, string>} headers The headers provided to the underlying WebSocket
     * @param {number} [timeout] Custom timeout value used per request . Defaults to `DEFAULT_TIMEOUT_MS`
     */
    constructor(endpoint = defaults.WS_URL, autoConnectMs = RETRY_DELAY, headers = {}, timeout) {
        _WsProvider_instances.add(this);
        _WsProvider_callCache.set(this, new LRUCache());
        _WsProvider_coder.set(this, void 0);
        _WsProvider_endpoints.set(this, void 0);
        _WsProvider_headers.set(this, void 0);
        _WsProvider_eventemitter.set(this, void 0);
        _WsProvider_handlers.set(this, {});
        _WsProvider_isReadyPromise.set(this, void 0);
        _WsProvider_stats.set(this, void 0);
        _WsProvider_waitingForId.set(this, {});
        _WsProvider_autoConnectMs.set(this, void 0);
        _WsProvider_endpointIndex.set(this, void 0);
        _WsProvider_endpointStats.set(this, void 0);
        _WsProvider_isConnected.set(this, false);
        _WsProvider_subscriptions.set(this, {});
        _WsProvider_timeoutId.set(this, null);
        _WsProvider_websocket.set(this, void 0);
        _WsProvider_timeout.set(this, void 0);
        _WsProvider_emit.set(this, (type, ...args) => {
            __classPrivateFieldGet(this, _WsProvider_eventemitter, "f").emit(type, ...args);
        });
        _WsProvider_onSocketClose.set(this, (event) => {
            const error = new Error(`disconnected from ${this.endpoint}: ${event.code}:: ${event.reason || getWSErrorString(event.code)}`);
            if (__classPrivateFieldGet(this, _WsProvider_autoConnectMs, "f") > 0) {
                l.error(error.message);
            }
            __classPrivateFieldSet(this, _WsProvider_isConnected, false, "f");
            if (__classPrivateFieldGet(this, _WsProvider_websocket, "f")) {
                __classPrivateFieldGet(this, _WsProvider_websocket, "f").onclose = null;
                __classPrivateFieldGet(this, _WsProvider_websocket, "f").onerror = null;
                __classPrivateFieldGet(this, _WsProvider_websocket, "f").onmessage = null;
                __classPrivateFieldGet(this, _WsProvider_websocket, "f").onopen = null;
                __classPrivateFieldSet(this, _WsProvider_websocket, null, "f");
            }
            if (__classPrivateFieldGet(this, _WsProvider_timeoutId, "f")) {
                clearInterval(__classPrivateFieldGet(this, _WsProvider_timeoutId, "f"));
                __classPrivateFieldSet(this, _WsProvider_timeoutId, null, "f");
            }
            // reject all hanging requests
            eraseRecord(__classPrivateFieldGet(this, _WsProvider_handlers, "f"), (h) => {
                try {
                    h.callback(error, undefined);
                }
                catch (err) {
                    // does not throw
                    l.error(err);
                }
            });
            eraseRecord(__classPrivateFieldGet(this, _WsProvider_waitingForId, "f"));
            // Reset stats for active endpoint
            __classPrivateFieldSet(this, _WsProvider_endpointStats, defaultEndpointStats(), "f");
            __classPrivateFieldGet(this, _WsProvider_emit, "f").call(this, 'disconnected');
            if (__classPrivateFieldGet(this, _WsProvider_autoConnectMs, "f") > 0) {
                setTimeout(() => {
                    this.connectWithRetry().catch(() => {
                        // does not throw
                    });
                }, __classPrivateFieldGet(this, _WsProvider_autoConnectMs, "f"));
            }
        });
        _WsProvider_onSocketError.set(this, (error) => {
            l.debug(() => ['socket error', error]);
            __classPrivateFieldGet(this, _WsProvider_emit, "f").call(this, 'error', error);
        });
        _WsProvider_onSocketMessage.set(this, (message) => {
            l.debug(() => ['received', message.data]);
            __classPrivateFieldGet(this, _WsProvider_endpointStats, "f").bytesRecv += message.data.length;
            __classPrivateFieldGet(this, _WsProvider_stats, "f").total.bytesRecv += message.data.length;
            const response = JSON.parse(message.data);
            return isUndefined(response.method)
                ? __classPrivateFieldGet(this, _WsProvider_onSocketMessageResult, "f").call(this, response)
                : __classPrivateFieldGet(this, _WsProvider_onSocketMessageSubscribe, "f").call(this, response);
        });
        _WsProvider_onSocketMessageResult.set(this, (response) => {
            const handler = __classPrivateFieldGet(this, _WsProvider_handlers, "f")[response.id];
            if (!handler) {
                l.debug(() => `Unable to find handler for id=${response.id}`);
                return;
            }
            try {
                const { method, params, subscription } = handler;
                const result = __classPrivateFieldGet(this, _WsProvider_coder, "f").decodeResponse(response);
                // first send the result - in case of subs, we may have an update
                // immediately if we have some queued results already
                handler.callback(null, result);
                if (subscription) {
                    const subId = `${subscription.type}::${result}`;
                    __classPrivateFieldGet(this, _WsProvider_subscriptions, "f")[subId] = objectSpread({}, subscription, {
                        method,
                        params
                    });
                    // if we have a result waiting for this subscription already
                    if (__classPrivateFieldGet(this, _WsProvider_waitingForId, "f")[subId]) {
                        __classPrivateFieldGet(this, _WsProvider_onSocketMessageSubscribe, "f").call(this, __classPrivateFieldGet(this, _WsProvider_waitingForId, "f")[subId]);
                    }
                }
            }
            catch (error) {
                __classPrivateFieldGet(this, _WsProvider_endpointStats, "f").errors++;
                __classPrivateFieldGet(this, _WsProvider_stats, "f").total.errors++;
                handler.callback(error, undefined);
            }
            delete __classPrivateFieldGet(this, _WsProvider_handlers, "f")[response.id];
        });
        _WsProvider_onSocketMessageSubscribe.set(this, (response) => {
            const method = ALIASES[response.method] || response.method || 'invalid';
            const subId = `${method}::${response.params.subscription}`;
            const handler = __classPrivateFieldGet(this, _WsProvider_subscriptions, "f")[subId];
            if (!handler) {
                // store the JSON, we could have out-of-order subid coming in
                __classPrivateFieldGet(this, _WsProvider_waitingForId, "f")[subId] = response;
                l.debug(() => `Unable to find handler for subscription=${subId}`);
                return;
            }
            // housekeeping
            delete __classPrivateFieldGet(this, _WsProvider_waitingForId, "f")[subId];
            try {
                const result = __classPrivateFieldGet(this, _WsProvider_coder, "f").decodeResponse(response);
                handler.callback(null, result);
            }
            catch (error) {
                __classPrivateFieldGet(this, _WsProvider_endpointStats, "f").errors++;
                __classPrivateFieldGet(this, _WsProvider_stats, "f").total.errors++;
                handler.callback(error, undefined);
            }
        });
        _WsProvider_onSocketOpen.set(this, () => {
            if (__classPrivateFieldGet(this, _WsProvider_websocket, "f") === null) {
                throw new Error('WebSocket cannot be null in onOpen');
            }
            l.debug(() => ['connected to', this.endpoint]);
            __classPrivateFieldSet(this, _WsProvider_isConnected, true, "f");
            __classPrivateFieldGet(this, _WsProvider_resubscribe, "f").call(this);
            __classPrivateFieldGet(this, _WsProvider_emit, "f").call(this, 'connected');
            return true;
        });
        _WsProvider_resubscribe.set(this, () => {
            const subscriptions = __classPrivateFieldGet(this, _WsProvider_subscriptions, "f");
            __classPrivateFieldSet(this, _WsProvider_subscriptions, {}, "f");
            Promise.all(Object.keys(subscriptions).map(async (id) => {
                const { callback, method, params, type } = subscriptions[id];
                // only re-create subscriptions which are not in author (only area where
                // transactions are created, i.e. submissions such as 'author_submitAndWatchExtrinsic'
                // are not included (and will not be re-broadcast)
                if (type.startsWith('author_')) {
                    return;
                }
                try {
                    await this.subscribe(type, method, params, callback);
                }
                catch (error) {
                    l.error(error);
                }
            })).catch(l.error);
        });
        _WsProvider_timeoutHandlers.set(this, () => {
            const now = Date.now();
            const ids = Object.keys(__classPrivateFieldGet(this, _WsProvider_handlers, "f"));
            for (let i = 0; i < ids.length; i++) {
                const handler = __classPrivateFieldGet(this, _WsProvider_handlers, "f")[ids[i]];
                if ((now - handler.start) > __classPrivateFieldGet(this, _WsProvider_timeout, "f")) {
                    try {
                        handler.callback(new Error(`No response received from RPC endpoint in ${__classPrivateFieldGet(this, _WsProvider_timeout, "f") / 1000}s`), undefined);
                    }
                    catch {
                        // ignore
                    }
                    __classPrivateFieldGet(this, _WsProvider_endpointStats, "f").timeout++;
                    __classPrivateFieldGet(this, _WsProvider_stats, "f").total.timeout++;
                    delete __classPrivateFieldGet(this, _WsProvider_handlers, "f")[ids[i]];
                }
            }
        });
        const endpoints = Array.isArray(endpoint)
            ? endpoint
            : [endpoint];
        if (endpoints.length === 0) {
            throw new Error('WsProvider requires at least one Endpoint');
        }
        endpoints.forEach((endpoint) => {
            if (!/^(wss|ws):\/\//.test(endpoint)) {
                throw new Error(`Endpoint should start with 'ws://', received '${endpoint}'`);
            }
        });
        __classPrivateFieldSet(this, _WsProvider_eventemitter, new EventEmitter(), "f");
        __classPrivateFieldSet(this, _WsProvider_autoConnectMs, autoConnectMs || 0, "f");
        __classPrivateFieldSet(this, _WsProvider_coder, new RpcCoder(), "f");
        __classPrivateFieldSet(this, _WsProvider_endpointIndex, -1, "f");
        __classPrivateFieldSet(this, _WsProvider_endpoints, endpoints, "f");
        __classPrivateFieldSet(this, _WsProvider_headers, headers, "f");
        __classPrivateFieldSet(this, _WsProvider_websocket, null, "f");
        __classPrivateFieldSet(this, _WsProvider_stats, {
            active: { requests: 0, subscriptions: 0 },
            total: defaultEndpointStats()
        }, "f");
        __classPrivateFieldSet(this, _WsProvider_endpointStats, defaultEndpointStats(), "f");
        __classPrivateFieldSet(this, _WsProvider_timeout, timeout || DEFAULT_TIMEOUT_MS, "f");
        if (autoConnectMs > 0) {
            this.connectWithRetry().catch(() => {
                // does not throw
            });
        }
        __classPrivateFieldSet(this, _WsProvider_isReadyPromise, new Promise((resolve) => {
            __classPrivateFieldGet(this, _WsProvider_eventemitter, "f").once('connected', () => {
                resolve(this);
            });
        }), "f");
    }
    /**
     * @summary `true` when this provider supports subscriptions
     */
    get hasSubscriptions() {
        return true;
    }
    /**
     * @summary `true` when this provider supports clone()
     */
    get isClonable() {
        return true;
    }
    /**
     * @summary Whether the node is connected or not.
     * @return {boolean} true if connected
     */
    get isConnected() {
        return __classPrivateFieldGet(this, _WsProvider_isConnected, "f");
    }
    /**
     * @description Promise that resolves the first time we are connected and loaded
     */
    get isReady() {
        return __classPrivateFieldGet(this, _WsProvider_isReadyPromise, "f");
    }
    get endpoint() {
        return __classPrivateFieldGet(this, _WsProvider_endpoints, "f")[__classPrivateFieldGet(this, _WsProvider_endpointIndex, "f")];
    }
    /**
     * @description Returns a clone of the object
     */
    clone() {
        return new WsProvider(__classPrivateFieldGet(this, _WsProvider_endpoints, "f"));
    }
    selectEndpointIndex(endpoints) {
        return (__classPrivateFieldGet(this, _WsProvider_endpointIndex, "f") + 1) % endpoints.length;
    }
    /**
     * @summary Manually connect
     * @description The [[WsProvider]] connects automatically by default, however if you decided otherwise, you may
     * connect manually using this method.
     */
    // eslint-disable-next-line @typescript-eslint/require-await
    async connect() {
        if (__classPrivateFieldGet(this, _WsProvider_websocket, "f")) {
            throw new Error('WebSocket is already connected');
        }
        try {
            __classPrivateFieldSet(this, _WsProvider_endpointIndex, this.selectEndpointIndex(__classPrivateFieldGet(this, _WsProvider_endpoints, "f")), "f");
            // the as typeof WebSocket here is Deno-specific - not available on the globalThis
            __classPrivateFieldSet(this, _WsProvider_websocket, typeof xglobal.WebSocket !== 'undefined' && isChildClass(xglobal.WebSocket, WebSocket)
                ? new WebSocket(this.endpoint)
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore - WS may be an instance of ws, which supports options
                : new WebSocket(this.endpoint, undefined, {
                    headers: __classPrivateFieldGet(this, _WsProvider_headers, "f")
                }), "f");
            if (__classPrivateFieldGet(this, _WsProvider_websocket, "f")) {
                __classPrivateFieldGet(this, _WsProvider_websocket, "f").onclose = __classPrivateFieldGet(this, _WsProvider_onSocketClose, "f");
                __classPrivateFieldGet(this, _WsProvider_websocket, "f").onerror = __classPrivateFieldGet(this, _WsProvider_onSocketError, "f");
                __classPrivateFieldGet(this, _WsProvider_websocket, "f").onmessage = __classPrivateFieldGet(this, _WsProvider_onSocketMessage, "f");
                __classPrivateFieldGet(this, _WsProvider_websocket, "f").onopen = __classPrivateFieldGet(this, _WsProvider_onSocketOpen, "f");
            }
            // timeout any handlers that have not had a response
            __classPrivateFieldSet(this, _WsProvider_timeoutId, setInterval(() => __classPrivateFieldGet(this, _WsProvider_timeoutHandlers, "f").call(this), TIMEOUT_INTERVAL), "f");
        }
        catch (error) {
            l.error(error);
            __classPrivateFieldGet(this, _WsProvider_emit, "f").call(this, 'error', error);
            throw error;
        }
    }
    /**
     * @description Connect, never throwing an error, but rather forcing a retry
     */
    async connectWithRetry() {
        if (__classPrivateFieldGet(this, _WsProvider_autoConnectMs, "f") > 0) {
            try {
                await this.connect();
            }
            catch (error) {
                setTimeout(() => {
                    this.connectWithRetry().catch(() => {
                        // does not throw
                    });
                }, __classPrivateFieldGet(this, _WsProvider_autoConnectMs, "f"));
            }
        }
    }
    /**
     * @description Manually disconnect from the connection, clearing auto-connect logic
     */
    // eslint-disable-next-line @typescript-eslint/require-await
    async disconnect() {
        // switch off autoConnect, we are in manual mode now
        __classPrivateFieldSet(this, _WsProvider_autoConnectMs, 0, "f");
        try {
            if (__classPrivateFieldGet(this, _WsProvider_websocket, "f")) {
                // 1000 - Normal closure; the connection successfully completed
                __classPrivateFieldGet(this, _WsProvider_websocket, "f").close(1000);
            }
        }
        catch (error) {
            l.error(error);
            __classPrivateFieldGet(this, _WsProvider_emit, "f").call(this, 'error', error);
            throw error;
        }
    }
    /**
     * @description Returns the connection stats
     */
    get stats() {
        return {
            active: {
                requests: Object.keys(__classPrivateFieldGet(this, _WsProvider_handlers, "f")).length,
                subscriptions: Object.keys(__classPrivateFieldGet(this, _WsProvider_subscriptions, "f")).length
            },
            total: __classPrivateFieldGet(this, _WsProvider_stats, "f").total
        };
    }
    get endpointStats() {
        return __classPrivateFieldGet(this, _WsProvider_endpointStats, "f");
    }
    /**
     * @summary Listens on events after having subscribed using the [[subscribe]] function.
     * @param  {ProviderInterfaceEmitted} type Event
     * @param  {ProviderInterfaceEmitCb}  sub  Callback
     * @return unsubscribe function
     */
    on(type, sub) {
        __classPrivateFieldGet(this, _WsProvider_eventemitter, "f").on(type, sub);
        return () => {
            __classPrivateFieldGet(this, _WsProvider_eventemitter, "f").removeListener(type, sub);
        };
    }
    /**
     * @summary Send JSON data using WebSockets to configured HTTP Endpoint or queue.
     * @param method The RPC methods to execute
     * @param params Encoded parameters as applicable for the method
     * @param subscription Subscription details (internally used)
     */
    send(method, params, isCacheable, subscription) {
        __classPrivateFieldGet(this, _WsProvider_endpointStats, "f").requests++;
        __classPrivateFieldGet(this, _WsProvider_stats, "f").total.requests++;
        const [id, body] = __classPrivateFieldGet(this, _WsProvider_coder, "f").encodeJson(method, params);
        let resultPromise = isCacheable
            ? __classPrivateFieldGet(this, _WsProvider_callCache, "f").get(body)
            : null;
        if (!resultPromise) {
            resultPromise = __classPrivateFieldGet(this, _WsProvider_instances, "m", _WsProvider_send).call(this, id, body, method, params, subscription);
            if (isCacheable) {
                __classPrivateFieldGet(this, _WsProvider_callCache, "f").set(body, resultPromise);
            }
        }
        else {
            __classPrivateFieldGet(this, _WsProvider_endpointStats, "f").cached++;
            __classPrivateFieldGet(this, _WsProvider_stats, "f").total.cached++;
        }
        return resultPromise;
    }
    /**
     * @name subscribe
     * @summary Allows subscribing to a specific event.
     *
     * @example
     * <BR>
     *
     * ```javascript
     * const provider = new WsProvider('ws://127.0.0.1:9944');
     * const rpc = new Rpc(provider);
     *
     * rpc.state.subscribeStorage([[storage.system.account, <Address>]], (_, values) => {
     *   console.log(values)
     * }).then((subscriptionId) => {
     *   console.log('balance changes subscription id: ', subscriptionId)
     * })
     * ```
     */
    subscribe(type, method, params, callback) {
        __classPrivateFieldGet(this, _WsProvider_endpointStats, "f").subscriptions++;
        __classPrivateFieldGet(this, _WsProvider_stats, "f").total.subscriptions++;
        // subscriptions are not cached, LRU applies to .at(<blockHash>) only
        return this.send(method, params, false, { callback, type });
    }
    /**
     * @summary Allows unsubscribing to subscriptions made with [[subscribe]].
     */
    async unsubscribe(type, method, id) {
        const subscription = `${type}::${id}`;
        // FIXME This now could happen with re-subscriptions. The issue is that with a re-sub
        // the assigned id now does not match what the API user originally received. It has
        // a slight complication in solving - since we cannot rely on the send id, but rather
        // need to find the actual subscription id to map it
        if (isUndefined(__classPrivateFieldGet(this, _WsProvider_subscriptions, "f")[subscription])) {
            l.debug(() => `Unable to find active subscription=${subscription}`);
            return false;
        }
        delete __classPrivateFieldGet(this, _WsProvider_subscriptions, "f")[subscription];
        try {
            return this.isConnected && !isNull(__classPrivateFieldGet(this, _WsProvider_websocket, "f"))
                ? this.send(method, [id])
                : true;
        }
        catch (error) {
            return false;
        }
    }
}
_WsProvider_callCache = new WeakMap(), _WsProvider_coder = new WeakMap(), _WsProvider_endpoints = new WeakMap(), _WsProvider_headers = new WeakMap(), _WsProvider_eventemitter = new WeakMap(), _WsProvider_handlers = new WeakMap(), _WsProvider_isReadyPromise = new WeakMap(), _WsProvider_stats = new WeakMap(), _WsProvider_waitingForId = new WeakMap(), _WsProvider_autoConnectMs = new WeakMap(), _WsProvider_endpointIndex = new WeakMap(), _WsProvider_endpointStats = new WeakMap(), _WsProvider_isConnected = new WeakMap(), _WsProvider_subscriptions = new WeakMap(), _WsProvider_timeoutId = new WeakMap(), _WsProvider_websocket = new WeakMap(), _WsProvider_timeout = new WeakMap(), _WsProvider_emit = new WeakMap(), _WsProvider_onSocketClose = new WeakMap(), _WsProvider_onSocketError = new WeakMap(), _WsProvider_onSocketMessage = new WeakMap(), _WsProvider_onSocketMessageResult = new WeakMap(), _WsProvider_onSocketMessageSubscribe = new WeakMap(), _WsProvider_onSocketOpen = new WeakMap(), _WsProvider_resubscribe = new WeakMap(), _WsProvider_timeoutHandlers = new WeakMap(), _WsProvider_instances = new WeakSet(), _WsProvider_send = async function _WsProvider_send(id, body, method, params, subscription) {
    return new Promise((resolve, reject) => {
        try {
            if (!this.isConnected || __classPrivateFieldGet(this, _WsProvider_websocket, "f") === null) {
                throw new Error('WebSocket is not connected');
            }
            const callback = (error, result) => {
                error
                    ? reject(error)
                    : resolve(result);
            };
            l.debug(() => ['calling', method, body]);
            __classPrivateFieldGet(this, _WsProvider_handlers, "f")[id] = {
                callback,
                method,
                params,
                start: Date.now(),
                subscription
            };
            __classPrivateFieldGet(this, _WsProvider_endpointStats, "f").bytesSent += body.length;
            __classPrivateFieldGet(this, _WsProvider_stats, "f").total.bytesSent += body.length;
            __classPrivateFieldGet(this, _WsProvider_websocket, "f").send(body);
        }
        catch (error) {
            __classPrivateFieldGet(this, _WsProvider_endpointStats, "f").errors++;
            __classPrivateFieldGet(this, _WsProvider_stats, "f").total.errors++;
            reject(error);
        }
    });
};
