// @polkadot/types-codec/cjs/extended/Range.js
"use strict";
var _Range_rangeName;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Range = void 0;
const tslib_1 = require("tslib");
const Tuple_js_1 = require("../base/Tuple.js");
/**
 * @name Range
 * @description
 * Rust `Range<T>` representation
 */
class Range extends Tuple_js_1.Tuple {
    constructor(registry, Type, value, { rangeName = 'Range' } = {}) {
        super(registry, [Type, Type], value);
        _Range_rangeName.set(this, void 0);
        tslib_1.__classPrivateFieldSet(this, _Range_rangeName, rangeName, "f");
    }
    static with(Type) {
        return class extends Range {
            constructor(registry, value) {
                super(registry, Type, value);
            }
        };
    }
    /**
     * @description Returns the starting range value
     */
    get start() {
        return this[0];
    }
    /**
     * @description Returns the ending range value
     */
    get end() {
        return this[1];
    }
    /**
     * @description Returns the base runtime type name for this instance
     */
    toRawType() {
        return `${tslib_1.__classPrivateFieldGet(this, _Range_rangeName, "f")}<${this.start.toRawType()}>`;
    }
}
exports.Range = Range;
_Range_rangeName = new WeakMap();
