// @polkadot/types-known/cjs/spec/rococo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const types_create_1 = require("@polkadot/types-create");
const util_1 = require("@polkadot/util");
/* eslint-disable sort-keys */
const sharedTypes = {
    DispatchErrorModule: 'DispatchErrorModuleU8',
    FullIdentification: '()',
    Keys: 'SessionKeys7B',
    Weight: 'WeightV1'
};
const versioned = [
    {
        minmax: [0, 200],
        types: (0, util_1.objectSpread)({}, sharedTypes, {
            AccountInfo: 'AccountInfoWithDualRefCount',
            Address: 'AccountId',
            LookupSource: 'AccountId'
        })
    },
    {
        minmax: [201, 214],
        types: (0, util_1.objectSpread)({}, sharedTypes, {
            AccountInfo: 'AccountInfoWithDualRefCount'
        })
    },
    {
        minmax: [215, 228],
        types: (0, util_1.objectSpread)({}, sharedTypes, {
            Keys: 'SessionKeys6'
        })
    },
    {
        minmax: [229, 9099],
        types: (0, util_1.objectSpread)({}, sharedTypes, (0, types_create_1.mapXcmTypes)('V0'))
    },
    {
        minmax: [9100, 9105],
        types: (0, util_1.objectSpread)({}, sharedTypes, (0, types_create_1.mapXcmTypes)('V1'))
    },
    {
        // metadata v14
        minmax: [9106, undefined],
        types: {
            Weight: 'WeightV1'
        }
    }
    // ,
    // {
    //   // weight v2 introduction
    //   minmax: [9300, undefined],
    //   types: {
    //     Weight: 'WeightV2'
    //   }
    // }
];
exports.default = versioned;
